### PR TITLE
RATIS-1150. Return DataStreamException to client

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Return DataStreamException to client

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1150

## How was this patch tested?

new u

### CODE REVIEW

**Comment by runzhiwang** [comment]
@szetszwo Could you help review this patch ? Thanks a lot.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java [hunk 1]
```diff
@@ -412,7 +417,11 @@ void read(DataStreamRequestByteBuf request, ChannelHandlerContext ctx,
           }
           buf.release();
           return null;
-        }, executor));
+        }, executor)).exceptionally(exception -> {
+      replyDataStreamException(server, exception, info.getRequest(), request, ctx);
+      buf.release();
+      return null;
+    });
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java:424]: Use whenComplete(..) and also a try-catch as below:
```suggestion
        }, executor)
    ).whenComplete((v, exception) -> {
      try {
        if (exception != null) {
          replyDataStreamException(server, exception, info.getRequest(), request, ctx);
        }
      } finally {
        buf.release();
      }
    });
```

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java [hunk 2]
```diff
@@ -412,7 +417,11 @@ void read(DataStreamRequestByteBuf request, ChannelHandlerContext ctx,
           }
           buf.release();
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java:418]: Remove `buf.release();` and use `whenComplete(..)` below.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java [hunk 3]
```diff
@@ -371,7 +373,10 @@ void read(DataStreamRequestByteBuf request, ChannelHandlerContext ctx,
       // for peers to start transaction
       final StreamInfo info = streams.get(key);
       composeAsync(info.getPrevious(), executor, v -> startTransaction(info, request, ctx))
-          .thenAccept(v -> buf.release());
+          .thenAccept(v -> buf.release()).exceptionally(exception -> {
+        replyDataStreamException(server, exception, info.getRequest(), request, ctx);
+        return null;
+      });
```
Comment (szetszwo) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java:379]: Use whenComplete(..) and also a try-catch as below:
```suggestion
          .whenComplete((v, exception) -> {
        try {
          if (exception != null) {
            replyDataStreamException(server, exception, info.getRequest(), request, ctx);
          }
        } finally {
          buf.release();
        }
      });
```

**Review by szetszwo — COMMENTED** [review]
The change looks good.  Just some minor comments inlined.

**Comment by runzhiwang** [comment]
@szetszwo Thanks for suggestions. I have updated the patch, could you help review it again ?

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**